File: Nursery/Stream/StreamDiscovery.swift
import Foundation

/// go2rtc knows its streams, so the app learns the detail and the everyday stream itself:
/// the stream with the most pixels is the detail (main) stream, the one with the fewest the
/// everyday (sub) stream. It runs in the test step of the guide. When nothing usable comes back,
/// the current names stay (`HomeDefaults`, or what the user typed).
enum StreamDiscovery {
    /// At most this many streams are asked, all at the same time.
    private static let limit = 8

    private struct Probe: Sendable {
        let name: String
        let width: Int
        let height: Int
        var pixels: Int { width * height }
        var text: String { "\(name) \(width)×\(height)" }
    }

    private static let session: URLSession = {
        let c = URLSessionConfiguration.ephemeral
        c.timeoutIntervalForRequest = 4
        c.waitsForConnectivity = false
        return URLSession(configuration: c)
    }()

    @MainActor static func run(settings: Settings) async {
        let host = settings.serverHost
        guard !host.isEmpty else { return }
        let names: [String]
        do {
            names = try await Self.streamNames(host: host)
        } catch {
            Log.shared.add("streams not read from go2rtc: \(error.localizedDescription)")
            return
        }
        // The current names first, so the limit never drops them. A name that is not safe in a
        // URL path cannot be played, so it is not asked.
        let current = [settings.streamMain, settings.streamSmall]
        func rank(_ name: String) -> Int { current.contains(name) ? 0 : 1 }
        let candidates = Array(names
            .filter { !$0.isEmpty && $0.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) == $0 }
            .sorted { rank($0) != rank($1) ? rank($0) < rank($1) : $0 < $1 }
            .prefix(limit))
        guard !candidates.isEmpty else {
            Log.shared.add("streams: go2rtc has none, keeping \(settings.streamMain), \(settings.streamSmall)")
            return
        }
        let found = await withTaskGroup(of: Probe?.self, returning: [String: Probe].self) { group in
            for name in candidates {
                group.addTask { await Self.probe(host: host, stream: name) }
            }
            var all: [String: Probe] = [:]
            for await p in group { if let p { all[p.name] = p } }
            return all
        }
        // In the order of the candidates, so a tie always picks the same stream.
        let usable = candidates.compactMap { found[$0] }
        let list = candidates.map { found[$0]?.text ?? "\($0) –" }.joined(separator: ", ")
        guard let detail = usable.max(by: { $0.pixels < $1.pixels }),
              let smallest = usable.min(by: { $0.pixels < $1.pixels }) else {
            Log.shared.add("streams: \(list) → no H.264 picture, keeping \(settings.streamMain), \(settings.streamSmall)")
            return
        }
        // Streams of the same size: one stream for both, so the app never switches for nothing.
        let everyday = smallest.pixels == detail.pixels ? detail : smallest
        Log.shared.add("streams: \(list) → detail \(detail.name), everyday \(everyday.name)")
        // Set only a change: each set redraws every view that watches the settings.
        if settings.streamMain != detail.name { settings.streamMain = detail.name }
        if settings.streamSmall != everyday.name { settings.streamSmall = everyday.name }
    }

    /// GET /api/streams: a JSON object, and its keys are the stream names.
    private static func streamNames(host: String) async throws -> [String] {
        guard let url = URL(string: "http://\(host):\(Go2rtc.apiPort)/api/streams") else { throw URLError(.badURL) }
        let (data, response) = try await session.data(from: url)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { throw URLError(.badServerResponse) }
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw URLError(.cannotParseResponse)
        }
        return Array(object.keys)
    }

    /// The picture size of one stream: from the SPS in the SDP, else from the first SPS in the stream.
    private static func probe(host: String, stream: String) async -> Probe? {
        let url = "rtsp://\(host):\(Go2rtc.rtspPort)/\(stream)"
        guard let client = try? RTSPClient(url: url) else { return nil }
        let timer = Task {
            try? await Task.sleep(for: .seconds(4))
            if !Task.isCancelled { client.stop() }
        }
        let tracks = try? await client.describe()
        timer.cancel()
        guard let video = tracks?.first(where: { $0.kind == .video && $0.codec == "H264" }) else { return nil }
        var sps = video.h264ParameterSets?.sps
        if sps == nil { sps = await Self.inBandSPS(url: url) }
        guard let sps, let size = H264SPS.size(sps) else { return nil }
        return Probe(name: stream, width: size.width, height: size.height)
    }

    /// The SDP has no SPS: play the stream until the first SPS comes, for 4 s at most.
    private static func inBandSPS(url: String) async -> [UInt8]? {
        guard let client = try? RTSPClient(url: url) else { return nil }
        let found = Found()
        let timer = Task {
            try? await Task.sleep(for: .seconds(4))
            if !Task.isCancelled { client.stop() }
        }
        defer { timer.cancel() }
        return await withCheckedContinuation { (cont: CheckedContinuation<[UInt8]?, Never>) in
            // It runs one time, for any end: the SPS came, the time is up, or an error.
            client.onClose = { _ in cont.resume(returning: found.sps) }
            Task {
                do {
                    _ = try await client.start { tracks in
                        guard let video = tracks.first(where: { $0.sdp.kind == .video && $0.sdp.codec == "H264" })?.channel else {
                            client.stop()
                            return
                        }
                        let depacketizer = H264Depacketizer()
                        client.onPacket = { channel, bytes in
                            guard channel == video, let packet = RTPPacket(bytes) else { return }
                            _ = depacketizer.push(packet)
                            if let sps = depacketizer.sps, found.sps == nil {
                                found.sps = sps
                                client.stop()
                            }
                        }
                    }
                } catch {
                    client.stop()
                }
            }
        }
    }

    /// The SPS from the RTSP queue.
    private final class Found: @unchecked Sendable {
        private let lock = NSLock()
        private var value: [UInt8]?
        var sps: [UInt8]? {
            get { lock.lock(); defer { lock.unlock() }; return value }
            set { lock.lock(); value = newValue; lock.unlock() }
        }
    }
}

File: android/app/src/main/java/cz/drabek/chuvicka/parent/StreamDiscovery.kt
package cz.drabek.chuvicka.parent

import cz.drabek.chuvicka.Go2rtc
import cz.drabek.chuvicka.Log
import cz.drabek.chuvicka.Settings
import cz.drabek.chuvicka.proto.H264Depacketizer
import cz.drabek.chuvicka.proto.H264Sps
import cz.drabek.chuvicka.proto.RtpPacket
import cz.drabek.chuvicka.proto.RtspClient
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * The app learns the two go2rtc streams itself: it asks go2rtc for its streams and reads the
 * picture size of each. The biggest is the detail stream, the smallest the everyday stream.
 * Blocking, for up to a few seconds per stream: call it off the main thread.
 */
object StreamDiscovery {
    private const val TIMEOUT = 4000
    private const val MAX_STREAMS = 8

    fun run(host: String) {
        if (host.isEmpty()) return
        val names = try { streamNames(host) } catch (e: Exception) {
            Log.add("streams: no list from go2rtc (${e.message}), keeping ${Settings.streamMain.value}, ${Settings.streamSmall.value}")
            return
        }
        val sizes = names.take(MAX_STREAMS).mapNotNull { name -> size(host, name)?.let { name to it } }
        if (sizes.isEmpty()) {
            Log.add("streams: none with a readable H.264 picture (${names.joinToString()}), keeping ${Settings.streamMain.value}, ${Settings.streamSmall.value}")
            return
        }
        val big = sizes.maxBy { it.second.first * it.second.second }
        val small = sizes.minBy { it.second.first * it.second.second }
        // The same size twice is one stream for both: a switch would give nothing.
        val everyday = if (small.second.first * small.second.second == big.second.first * big.second.second) big else small
        Log.add("streams: ${sizes.joinToString { "${it.first} ${it.second.first}×${it.second.second}" }} → detail ${big.first}, everyday ${everyday.first}")
        Settings.set(Settings.streamMain, "streamMain", big.first)
        Settings.set(Settings.streamSmall, "streamSmall", everyday.first)
    }

    /** GET /api/streams: a JSON object, its keys are the stream names. */
    private fun streamNames(host: String): List<String> {
        val c = URL("http://$host:${Go2rtc.API_PORT}/api/streams").openConnection() as HttpURLConnection
        c.connectTimeout = TIMEOUT
        c.readTimeout = TIMEOUT
        try {
            if (c.responseCode != 200) throw java.io.IOException("HTTP ${c.responseCode}")
            val json = JSONObject(c.inputStream.use { it.readBytes() }.toString(Charsets.UTF_8))
            return json.keys().asSequence().filter { it.isNotBlank() && '/' !in it && ' ' !in it }.toList()
        } finally {
            c.disconnect()
        }
    }

    /** The picture size: from the SDP, else from the first SPS in the stream (at most about 3 s). */
    private fun size(host: String, name: String): Pair<Int, Int>? {
        val client = RtspClient("rtsp://$host:${Go2rtc.RTSP_PORT}/$name", host, Go2rtc.RTSP_PORT, timeoutMs = TIMEOUT)
        var found: Pair<Int, Int>? = null
        var timer: Thread? = null
        try {
            val video = client.start().firstOrNull { it.sdp.kind == "video" } ?: return null
            video.sdp.h264ParameterSets?.let { (sps, _) -> H264Sps.size(sps) }?.let { return it }
            timer = Thread {
                try { Thread.sleep(3000) } catch (_: InterruptedException) {}
                client.close()
            }.apply { isDaemon = true; start() }
            val depacketizer = H264Depacketizer()
            client.play { channel, bytes ->
                if (channel != video.channel) return@play
                depacketizer.push(RtpPacket.parse(bytes) ?: return@play)
                val sps = depacketizer.sps ?: return@play
                found = H264Sps.size(sps)
                client.close()
            }
        } catch (e: Exception) {
            if (found == null) Log.add("streams: $name not read (${e.message})")
        } finally {
            timer?.interrupt()
            client.close()
        }
        return found
    }
}
